### PR TITLE
[SPARK-44030][SQL][FOLLOW-UP] Move unapply from AnyTimestampType to AnyTimestampTypeExpression

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Move unapply from AnyTimestampType to AnyTimestampTypeExpression.

### Why are the changes needed?

To align with the effort that we use separate type expression class to host `unapply`.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Existing Test

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell @cloud-fan

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 1]
```diff
@@ -294,14 +294,16 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
-      case d @ DateAdd(AnyTimestampType(), _) => d.copy(startDate = Cast(d.startDate, DateType))
-      case d @ DateSub(AnyTimestampType(), _) => d.copy(startDate = Cast(d.startDate, DateType))
+      case d @ DateAdd(AnyTimestampTypeExpression(), _)
+      => d.copy(startDate = Cast(d.startDate, DateType))
+      case d @ DateSub(AnyTimestampTypeExpression(), _)
+      => d.copy(startDate = Cast(d.startDate, DateType))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:300]: Could you change coding style here, please:
```suggestion
      case d @ DateAdd(AnyTimestampTypeExpression(), _) =>
        d.copy(startDate = Cast(d.startDate, DateType))
      case d @ DateSub(AnyTimestampTypeExpression(), _) =>
        d.copy(startDate = Cast(d.startDate, DateType))
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala]: done

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @amaliujia and @cloud-fan for review.

**Pull request closed without merging**